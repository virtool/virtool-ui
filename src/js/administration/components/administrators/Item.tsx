import React from "react";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../../app/theme";
import { BoxSpaced, Icon, InitialIcon } from "../../../base";
import { User } from "../../../users/types";
import { useSetAdministratorRole } from "../../queries";
import { AdministratorRoles } from "../../types";
import { RoleSelect } from "./RoleSelect";

const StyledAdministrator = styled(BoxSpaced)`
    display: flex;
    align-items: center;

    svg {
        margin-right: 10px;
    }

    button {
        margin-left: auto;
    }

    button:last-child {
        margin-left: 10px;
    }
`;

const UserHandle = styled.span`
    font-weight: ${getFontWeight("thick")};
    font-size: ${getFontSize("lg")};
`;

const InlineRoleSelect = styled(RoleSelect)`
    width: 200px;
`;

type AdministratorItemProps = {
    user: User;
    roles: Array<AdministratorRoles>;
};

export const AdministratorItem = ({ user, roles }: AdministratorItemProps) => {
    const editMutation = useSetAdministratorRole();
    const onChange = (value: AdministratorRoles) => {
        editMutation.mutate({ user_id: user.id, role: value });
    };

    return (
        <StyledAdministrator key={user.id}>
            <InitialIcon handle={user.handle} size="lg" />
            <UserHandle>{user.handle}</UserHandle>
            <InlineRoleSelect value={user.administrator_role} roles={roles} onChange={onChange} />
            <Icon name="trash" color="red" onClick={() => onChange(null)} aria-label="Remove administrator role" />
        </StyledAdministrator>
    );
};
