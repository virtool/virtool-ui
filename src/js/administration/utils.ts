import { Account } from "@account/types";
import { Permission } from "../groups/types";
import { AdministratorRoles } from "./types";

/**
 * The permissions level of each administrator role
 */
enum AdministratorPermissionsLevel {
    full,
    settings,
    spaces,
    users,
    base,
}

/**
 * Check if a user has a sufficient admin role
 *
 * @param requiredRole - The lowest admin role the user must have to pass the check
 * @param userRole - The administrator role of the user
 */

export function hasSufficientAdminRole(requiredRole: AdministratorRoles, userRole: AdministratorRoles): boolean {
    return AdministratorPermissionsLevel[userRole] <= AdministratorPermissionsLevel[requiredRole];
}

/**
 * Permissions granted to each administrator role
 */
export enum AdministratorPermissions {
    cancel_job = AdministratorRoles.BASE,
    create_ref = AdministratorRoles.BASE,
    modify_hmm = AdministratorRoles.BASE,
    remove_job = AdministratorRoles.BASE,
    upload_file = AdministratorRoles.FULL,
    create_sample = AdministratorRoles.FULL,
    modify_subtraction = AdministratorRoles.FULL,
    remove_file = AdministratorRoles.FULL,
}

/**
 * Check if a user has a sufficient admin role or legacy permissions to perform an action
 *
 * @param account - The Account object of the user
 * @param permission - The permissions to check
 * @returns  Whether the user is allowed to perform the action
 */
export function checkAdminRoleOrPermissionsFromAccount(account: Account, permission: Permission): boolean {
    return (
        hasSufficientAdminRole(AdministratorPermissions[permission as string], account.administrator_role) ||
        account.permissions[permission]
    );
}
