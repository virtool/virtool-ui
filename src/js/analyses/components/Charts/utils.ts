import { filter, reduce, zipWith } from "lodash-es";
import { IimiSequence, UntrustworthyRange } from "../../types";
/**
 * Convert range length encoded data into an array of depth values.
 *
 * @param lengths - the lengths of a given read depth region
 * @param rle - the depth values of the regions
 *
 * @returns An array of depth values
 */
export function convertRleToCoverage(lengths: Array<number>, rle: Array<number>) {
    const coverage = [];

    for (let sharedIndex = 0; sharedIndex < lengths.length; sharedIndex++) {
        const length = lengths[sharedIndex];
        const value = rle[sharedIndex];

        coverage.push(...Array(length).fill(value));
    }

    return coverage;
}

/**
 * Average the read depths of an array of sequences.
 *
 * @param sequences - the sequences to average
 * @returns An averaged sequence
 */
export function averageSequences(sequences: IimiSequence[]) {
    return zipWith(...sequences, (...args) => {
        const filtered = filter(args, num => typeof num === "number");
        return filtered.length === 0 ? 0 : reduce(filtered, (a, b) => a + b) / filtered.length;
    });
}

/**
 * Determine the regions of the sequence that are trustworthy using the untrustworthy regions.
 *
 * @param length - The length of the sequence
 * @param untrustworthyRanges - the untrustworthy regions to infer trustworthy regions from
 */
export function deriveTrustworthyRegions(length: number, untrustworthyRanges: [number, number][]): [number, number][] {
    const trustworthyRanges = [];
    let start = 1;
    let end;

    untrustworthyRanges.forEach(range => {
        end = range[0];
        trustworthyRanges.push([start, end]);
        start = range[1];
    });

    trustworthyRanges.push([start, length]);
    return trustworthyRanges;
}

/**
 * Construct a union of untrustworthy regions from multiple sources.
 *
 * @param untrustworthyRanges - A collection of untrustworthy regions to combine
 * @returns the combined untrustworthy regions
 */
export function combineUntrustworthyRegions(untrustworthyRanges: UntrustworthyRange[][]): UntrustworthyRange[] {
    const sortedUntrustworthyRanges = untrustworthyRanges.flat().sort((a, b) => a[0] - b[0]);

    if (sortedUntrustworthyRanges.length === 0) {
        return [];
    }

    const combined = [] as [number, number][];
    combined.push(sortedUntrustworthyRanges.shift());

    sortedUntrustworthyRanges.forEach(range => {
        const last = combined[combined.length - 1];

        if (range[0] <= last[1]) {
            last[1] = range[1];
        } else {
            combined.push(range);
        }
    });

    return combined;
}
