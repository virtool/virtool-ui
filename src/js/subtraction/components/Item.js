import React from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../app/theme";
import { Icon, LinkBox, Loader } from "../../base";

export const SubtractionItemIcon = ({ ready }) => {
    if (ready) {
        return <Icon name="check" color="green" />;
    }

    return <Loader size="14px" color="primary" />;
};

const StyledSubtractionItem = styled(LinkBox)`
    align-items: center;
    display: flex;
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};

    > span:last-child {
        margin-left: auto;
    }
`;

export const SubtractionItem = ({ id, name, ready }) => (
    <StyledSubtractionItem key={id} to={`/subtraction/${id}`}>
        <span>{name}</span>
        <span>
            <SubtractionItemIcon ready={ready} /> <span>{ready ? "Ready" : "Importing"}</span>
        </span>
    </StyledSubtractionItem>
);

export const mapStateToProps = (state, props) => {
    const { id, name, ready } = state.subtraction.documents[props.index];
    return {
        id,
        name,
        ready
    };
};

export default connect(mapStateToProps)(SubtractionItem);
