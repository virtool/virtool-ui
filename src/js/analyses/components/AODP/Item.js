import React, { useCallback } from "react";
import styled from "styled-components";
import { connect } from "react-redux";
import { Badge } from "../../../base";

import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";

const StyledAODPItem = styled(AnalysisViewerItem)`
    display: flex;
    justify-content: space-between;
`;

export const AODPItem = ({ active, hits, id, name, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(id), [id]);

    return (
        <StyledAODPItem active={active} style={style} onClick={handleClick}>
            <strong>{name}</strong>
            <Badge>{hits}</Badge>
        </StyledAODPItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).id;
    const matches = getMatches(state);
    const match = matches[ownProps.index];

    const { id, identities, name } = match;
    return { id, name, hits: identities.length, active: activeId === id };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AODPItem);
