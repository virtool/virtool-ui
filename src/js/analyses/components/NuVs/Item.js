import { filter } from "lodash-es";
import numbro from "numbro";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { Badge } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";
import { NuVsValues } from "./Values";

const calculateAnnotatedOrfCount = orfs => filter(orfs, orf => orf.hits.length).length;

const NuVsItemHeader = styled.div`
    align-items: center;
    display: flex;
    justify-content: space-between;
`;

const StyledNuVsItem = styled(AnalysisViewerItem)`
    border-bottom: none;
    border-left: none;
    border-radius: 0;
    margin: 0;
`;

export const NuVsItem = ({ active, e, orfs, sequence, sequenceIndex, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(sequenceIndex), [sequenceIndex]);

    return (
        <StyledNuVsItem active={active} onClick={handleClick} style={style}>
            <NuVsItemHeader>
                <strong>Sequence {sequenceIndex}</strong>
                <Badge>{sequence.length}</Badge>
            </NuVsItemHeader>
            <NuVsValues e={numbro(e).format()} orfCount={calculateAnnotatedOrfCount(orfs)} />
        </StyledNuVsItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).index;
    const { e, index, orfs, sequence } = getMatches(state)[ownProps.index];
    return { e, orfs, sequence, sequenceIndex: index, active: activeId === index };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NuVsItem);
