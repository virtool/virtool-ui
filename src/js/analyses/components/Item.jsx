import React from "react";
import { connect } from "react-redux";
import { Link, useRouteMatch } from "react-router-dom";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../app/theme";
import { Attribution, BoxSpaced, Icon, SlashList } from "../../base";
import { getCanModify } from "../../samples/selectors";
import { getWorkflowDisplayName } from "../../utils/utils";
import { removeAnalysis } from "../actions";
import { AnalysisItemRightIcon } from "./RightIcon";

const StyledAnalysisItem = styled(BoxSpaced)`
    color: ${props => props.theme.color.greyDarkest};

    &:hover {
        ${props => (props.ready ? "background-color: lightgrey;" : "")};
    }
`;

const AnalysisItemTag = styled.span`
    align-items: center;
    display: inline-flex;
    margin-right: 15px;

    ${SlashList} {
        margin: 0;
    }

    i {
        margin-right: 5px;
    }
`;

const AnalysisItemTags = styled.div`
    align-items: center;
    display: flex;
    margin-top: 10px;
`;

const AnalysisItemTop = styled.div`
    align-items: center;
    display: flex;
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};
    justify-content: space-between;

    a {
        font-weight: ${getFontWeight("thick")};
    }
`;

function AnalysisItem({ canModify, created_at, id, index, ready, reference, subtractions, user, workflow, onRemove }) {
    const sampleId = useRouteMatch().params.sampleId;

    return (
        <StyledAnalysisItem>
            <AnalysisItemTop>
                <Link to={`/samples/${sampleId}/analyses/${id}`}>{getWorkflowDisplayName(workflow)}</Link>
                <AnalysisItemRightIcon canModify={canModify} onRemove={onRemove} ready={ready} />
            </AnalysisItemTop>
            <Attribution user={user.handle} time={created_at} />
            <AnalysisItemTags>
                <AnalysisItemTag key="reference">
                    <Icon name="equals" />
                    <SlashList>
                        <li>
                            <Link to={`/refs/${reference.id}`}>{reference.name}</Link>
                        </li>
                        <li>
                            <Link to={`/refs/${reference.id}/indexes/${index.id}`}>Index {index.version}</Link>
                        </li>
                    </SlashList>
                </AnalysisItemTag>
                {subtractions.map(subtraction => (
                    <AnalysisItemTag key={subtraction.id}>
                        <Icon name="not-equal" />
                        <Link to={`/subtractions/${subtraction.id}`}>{subtraction.name}</Link>
                    </AnalysisItemTag>
                ))}
            </AnalysisItemTags>
        </StyledAnalysisItem>
    );
}

export function mapStateToProps(state) {
    return {
        canModify: getCanModify(state)
    };
}

function mapDispatchToProps(dispatch, ownProps) {
    return {
        onRemove: () => {
            dispatch(removeAnalysis(ownProps.id));
        }
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(AnalysisItem);
