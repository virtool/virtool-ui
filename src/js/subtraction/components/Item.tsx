import React from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { getFontSize, getFontWeight, sizes } from "../../app/theme";
import { BoxLink } from "../../base";
import { ProgressCircle } from "../../base/ProgressCircle";
import { JobMinimal } from "../../jobs/types";
import { getStateTitle } from "../../jobs/utils";
import { UserNested } from "../../users/types";
import { SubtractionAttribution } from "./Attribution";

const StyledSubtractionItemHeader = styled.div`
    align-items: center;
    display: flex;
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};

    > span:last-child {
        margin-left: auto;
    }
`;

const ProgressTag = styled.span`
    display: flex;
    align-items: center;
    svg {
        margin-right: 5px;
    }
`;

type SubtractionItemProps = {
    id: string;
    user: UserNested;
    name: string;
    created_at: string;
    job: JobMinimal;
};

/**
 * Condensed subtraction item for use in a list of subtractions
 *
 * @param id - The unique subtraction id
 * @param user - The user who created the subtraction
 * @param name - The name of the subtraction
 * @param created_at - The date the subtraction was created
 * @param job - The job associated with the subtraction
 * @returns A condensed subtraction item
 */
export function SubtractionItem({ id, user, name, created_at, job }: SubtractionItemProps) {
    return (
        <BoxLink key={id} to={`/subtractions/${id}`}>
            <StyledSubtractionItemHeader>
                <span>{name}</span>
                <ProgressTag>
                    {job.state === "complete" || (
                        <>
                            <ProgressCircle size={sizes.md} progress={job.progress} state={job.state} />
                            {getStateTitle(job.state)}
                        </>
                    )}
                </ProgressTag>
            </StyledSubtractionItemHeader>
            <SubtractionAttribution handle={user.handle} time={created_at} />
        </BoxLink>
    );
}

/**
 * Extract information from redux state to pass as props to SubtractionItem
 *
 * @param state - The complete current redux state
 * @param props - The props passed to the SubtractionItem component
 * @returns The props derived from redux state
 */
export function mapStateToProps(state, props) {
    const { id, user, name, created_at, job } = state.subtraction.documents[props.index];
    return {
        id,
        user,
        created_at,
        name,
        job,
    };
}

export default connect(mapStateToProps)(SubtractionItem);
