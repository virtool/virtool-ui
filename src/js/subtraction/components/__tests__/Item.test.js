import { SubtractionItem, SubtractionItemIcon, mapStateToProps } from "../Item";

describe("<SubtractionItemIcon />", () => {
    it.each([true, false])("should render when [ready=%p]", ready => {
        const wrapper = shallow(<SubtractionItemIcon ready={ready} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("<SubtractionItem />", () => {
    let props;

    beforeEach(() => {
        props = {
            id: "foo",
            name: "Foo",
            ready: true
        };
    });

    it.each([true, false])("should render when [ready=%p]", ready => {
        props.ready = ready;
        const wrapper = shallow(<SubtractionItem {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("mapStateToProps()", () => {
    it("should return props", () => {
        const state = {
            subtraction: {
                documents: [
                    { id: "foo", name: "Foo", ready: true },
                    { id: "bar", name: "Bar", ready: true },
                    { id: "baz", name: "Baz", ready: true }
                ]
            }
        };
        const props = mapStateToProps(state, { index: 1 });
        expect(props).toEqual({
            id: "bar",
            name: "Bar",
            ready: true
        });
    });
});
