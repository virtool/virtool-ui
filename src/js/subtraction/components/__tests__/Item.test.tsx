import { screen } from "@testing-library/react";
import { createBrowserHistory } from "history";
import React from "react";
import { beforeEach, describe, expect, it } from "vitest";
import { renderWithRouter } from "../../../../tests/setupTests";
import { getStateTitle } from "../../../jobs/utils";
import { mapStateToProps, SubtractionItem } from "../Item";

describe("<SubtractionItem />", () => {
    let props;
    let history;

    beforeEach(() => {
        props = {
            id: "foo",
            name: "Foo",
            user: { handle: "user_handle" },
            job: { progress: 50, state: "running" },
            created_at: new Date().setFullYear(new Date().getFullYear() - 1),
        };
        history = createBrowserHistory();
    });

    it("should render", () => {
        renderWithRouter(<SubtractionItem {...props} />, {}, history);
        expect(screen.getByText("Foo")).toBeInTheDocument();
        expect(screen.getByText(/user_handle/)).toBeInTheDocument();
        expect(screen.getByText(/1 year ago/)).toBeInTheDocument();
        expect(screen.getByRole("progressbar")).toHaveAttribute("data-value", "50");
        expect(screen.getByText("Running")).toBeInTheDocument();
    });

    it.each(["waiting", "running", "error"])("should render %s state", state => {
        props.job.state = state;
        renderWithRouter(<SubtractionItem {...props} />, {}, history);
        expect(screen.getByText(getStateTitle(state))).toBeInTheDocument();
    });

    it("should not render progress bar if job is complete", () => {
        props.job.state = "complete";
        renderWithRouter(<SubtractionItem {...props} />, {}, history);
        expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
        expect(screen.queryByText("Complete")).not.toBeInTheDocument();
    });
});

describe("mapStateToProps()", () => {
    it("should return props", () => {
        const state = {
            subtraction: {
                documents: [
                    {
                        id: "foo",
                        name: "Foo",
                        ready: true,
                        user: { id: "user_id_1", handle: "user_handle_1" },
                        job: { id: "job_id_1", progress: 50, state: "running" },
                    },
                    {
                        id: "bar",
                        name: "Bar",
                        ready: true,
                        user: { id: "user_id_2", handle: "user_handle_2" },
                        job: { id: "job_id_2", progress: 50, state: "failed" },
                    },
                    {
                        id: "baz",
                        name: "Baz",
                        ready: true,
                        user: { id: "user_id_3", handle: "user_handle_3" },
                        job: { id: "job_id_2", progress: 100, state: "complete" },
                    },
                ],
            },
        };
        const props = mapStateToProps(state, { index: 1 });
        expect(props).toEqual({
            id: "bar",
            name: "Bar",
            job: { id: "job_id_2", progress: 50, state: "failed" },
            user: { id: "user_id_2", handle: "user_handle_2" },
        });
    });
});
