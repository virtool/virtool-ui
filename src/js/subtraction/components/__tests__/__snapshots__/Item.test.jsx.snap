// Vitest Snapshot v1

exports[`<SubtractionItem /> > should render when [create_at=%p] 0 1`] = `
<BoxLink
  key="foo"
  to="/subtractions/foo"
>
  <styled.div>
    <span>
      Foo
    </span>
    <span>
      <SubtractionItemIcon
        ready={true}
      />
       
      Ready
    </span>
  </styled.div>
  <SubtractionAttribution
    handle="bar"
    time={0}
  />
</BoxLink>
`;

exports[`<SubtractionItem /> > should render when [create_at=%p] null 1`] = `
<BoxLink
  key="foo"
  to="/subtractions/foo"
>
  <styled.div>
    <span>
      Foo
    </span>
    <span>
      <SubtractionItemIcon
        ready={true}
      />
       
      Ready
    </span>
  </styled.div>
  <SubtractionAttribution
    handle="bar"
    time={null}
  />
</BoxLink>
`;

exports[`<SubtractionItem /> > should render when [ready=%p] false 1`] = `
<BoxLink
  key="foo"
  to="/subtractions/foo"
>
  <styled.div>
    <span>
      Foo
    </span>
    <span>
      <SubtractionItemIcon
        ready={false}
      />
       
      Creating
    </span>
  </styled.div>
  <SubtractionAttribution
    handle="bar"
  />
</BoxLink>
`;

exports[`<SubtractionItem /> > should render when [ready=%p] true 1`] = `
<BoxLink
  key="foo"
  to="/subtractions/foo"
>
  <styled.div>
    <span>
      Foo
    </span>
    <span>
      <SubtractionItemIcon
        ready={true}
      />
       
      Ready
    </span>
  </styled.div>
  <SubtractionAttribution
    handle="bar"
  />
</BoxLink>
`;

exports[`<SubtractionItemIcon /> > should render when [ready=%p] false 1`] = `
<Loader
  color="primary"
  size="14px"
/>
`;

exports[`<SubtractionItemIcon /> > should render when [ready=%p] true 1`] = `
<Icon
  color="green"
  name="check"
/>
`;
